stores/sql/metrics.go

```diff
@@ -0,0 +1,142 @@
+package sql
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/internal/sql"
+)
+
+const (
+	tableContractPruneMetric = "contract_prunes"
+)
+
+func ContractPruneMetrics(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts api.ContractPruneMetricsQueryOpts) (metrics []api.ContractPruneMetric, _ error) {
+	rows, err := queryPeriods(ctx, tx, start, n, interval, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch contract metrics: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var cpm api.ContractPruneMetric
+		var placeHolder int64
+		var placeHolderTime time.Time
+		var timestamp UnixTimeMS
+		if err := rows.Scan(&placeHolder, &placeHolderTime, &timestamp, (*FileContractID)(&cpm.ContractID), (*PublicKey)(&cpm.HostKey), &cpm.HostVersion, (*Unsigned64)(&cpm.Pruned), (*Unsigned64)(&cpm.Remaining), &cpm.Duration); err != nil {
+			return nil, fmt.Errorf("failed to scan contract prune metric: %w", err)
+		}
+		cpm.Timestamp = api.TimeRFC3339(timestamp)
+		metrics = append(metrics, cpm)
+	}
+
+	return metrics, nil
+}
+
+func RecordContractPruneMetric(ctx context.Context, tx sql.Tx, metrics ...api.ContractPruneMetric) error {
+	insertStmt, err := tx.Prepare(ctx, fmt.Sprintf("INSERT INTO %s (created_at, timestamp, fcid, host, host_version, pruned, remaining, duration) VALUES (?, ?,?, ?, ?, ?, ?, ?)", tableContractPruneMetric))
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert contract prune metric: %w", err)
+	}
+	defer insertStmt.Close()
+
+	for _, metric := range metrics {
+		res, err := insertStmt.Exec(ctx,
+			time.Now().UTC(),
+			UnixTimeMS(metric.Timestamp),
+			FileContractID(metric.ContractID),
+			PublicKey(metric.HostKey),
+			metric.HostVersion,
+			Unsigned64(metric.Pruned),
+			Unsigned64(metric.Remaining),
+			metric.Duration,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert contract prune metric: %w", err)
+		} else if n, err := res.RowsAffected(); err != nil {
+			return fmt.Errorf("failed to get rows affected: %w", err)
+		} else if n == 0 {
+			return fmt.Errorf("failed to insert contract prune metric: no rows affected")
+		}
+	}
+
+	return nil
+}
+
+func queryPeriods(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts interface{}) (*sql.LoggedRows, error) {
+	if n > api.MetricMaxIntervals {
+		return nil, api.ErrMaxIntervalsExceeded
+	}
+
+	params := []interface{}{
+		UnixTimeMS(start),
+		interval.Milliseconds(),
+		UnixTimeMS(start.Add(time.Duration(n) * interval)),
+		interval.Milliseconds(),
+		interval.Milliseconds(),
+	}
+
+	where, err := whereClauseFromQueryOpts(opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build where clause: %w", err)
+	} else if len(where.params) > 0 {
+		params = append(params, where.params...)
+	}
+
+	return tx.Query(ctx, fmt.Sprintf(`
+	WITH RECURSIVE periods AS (
```

Comment:
I think the other SQL methods add one more tab before the query to not align the start of the query with the go code directly.